system-parachains/collectives/collectives-polkadot/src/ambassador/mod.rs

```diff
@@ -0,0 +1,358 @@
+//! (via [AmbassadorCore](pallet_core_fellowship)).
+//! - Members' salaries (via [AmbassadorSalary](pallet_salary), requiring a member to be
+//! imported or inducted into [AmbassadorCore](pallet_core_fellowship)).
+//! - Ambassador Program Sub-Treasury (via [AmbassadorTreasury](pallet_treasury)).
+
+pub mod origins;
+mod tracks;
+
+pub use origins::pallet_origins as pallet_ambassador_origins;
+
+use crate::{
+	xcm_config::{AssetHubUsdt, FellowshipAdminBodyId},
+	*,
+};
+use frame_support::{
+	pallet_prelude::PalletInfoAccess,
+	traits::{EitherOf, MapSuccess, TryMapSuccess},
+};
+use frame_system::EnsureRootWithSuccess;
+use origins::pallet_origins::{EnsureAmbassadorsFrom, HeadAmbassadors, Origin, SeniorAmbassadors};
+use pallet_ranked_collective::{Rank, Votes};
+use polkadot_runtime_common::impls::{LocatableAssetConverter, VersionedLocationConverter};
+use sp_core::ConstU128;
+use sp_runtime::{
+	traits::{CheckedReduceBy, Convert, ConvertToValue, IdentityLookup, Replace},
+	Permill,
+};
+use xcm::prelude::*;
+use xcm_builder::{AliasesIntoAccountId32, PayOverXcm};
+
+/// The Ambassador Program's member ranks.
+pub mod ranks {
+	use super::Rank;
+
+	#[allow(dead_code)]
+	pub const CANDIDATE: Rank = 0;
+	pub const AMBASSADOR: Rank = 1;
+	pub const SENIOR_AMBASSADOR: Rank = 2;
+	pub const HEAD_AMBASSADOR: Rank = 3;
+}
+
+impl pallet_ambassador_origins::Config for Runtime {}
+
+/// Demotion is by any of:
+/// - Root can demote arbitrarily;
+/// - the FellowshipAdmin voice (i.e. token holder referendum) can demote arbitrarily;
+/// - Head Ambassadors voice can demote Senior Ambassador or Ambassador;
+/// - Senior Ambassadors voice can demote Ambassador.
+pub type DemoteOrigin = EitherOf<
+	EnsureRootWithSuccess<AccountId, ConstU16<65535>>,
+	EitherOf<
+		MapSuccess<
+			EnsureXcm<IsVoiceOfBody<GovernanceLocation, FellowshipAdminBodyId>>,
+			Replace<ConstU16<{ ranks::HEAD_AMBASSADOR }>>,
+		>,
+		TryMapSuccess<
+			EnsureAmbassadorsFrom<ConstU16<{ ranks::SENIOR_AMBASSADOR }>>,
+			CheckedReduceBy<ConstU16<1>>,
+		>,
+	>,
+>;
+
+/// Promotion and approval (rank-retention) is by any of:
+/// - Root can promote arbitrarily.
+/// - the FellowshipAdmin voice (i.e. token holder referendum) can promote arbitrarily.
+/// - Head Ambassadors voice can promote to Senior Ambassador and Ambassador;
+/// - Senior Ambassadors voice can promote to Ambassador.
+pub type PromoteOrigin = DemoteOrigin;
+
+/// Root, FellowshipAdmin or HeadAmbassadors.
+pub type OpenGovOrHeadAmbassadors = EitherOfDiverse<
+	EnsureRoot<AccountId>,
+	EitherOfDiverse<
+		HeadAmbassadors,
+		EnsureXcm<IsVoiceOfBody<GovernanceLocation, FellowshipAdminBodyId>>,
+	>,
+>;
+
+/// Ambassadors' vote weights for referendums.
+/// - Each member with an excess rank of 0 gets 1 vote;
+/// - ...with an excess rank of 1 gets 5 votes;
+/// - ...with an excess rank of 2 gets 10 votes;
+/// - ...with an excess rank of 3 gets 15 votes;
```

Comment:
there is no way with a current version of the pallet to describe the weights for the votes the way it is manifested (but it is possible if we update this convert contract in the pallet).
here I keep the proportion for SA and HA as in the document, but A gets less.

Reply:
Interesting, thought it would be possible to basically hardcode a `rank => weight` table. I think the intent of the manifesto is that HA should be **much** more powerful because they can be highly outnumbered but are selected by general referendum. As in:

- Can only be 21 HAs, but they are selected by all Polkadot holders,
- Can be potentially 10s or 100s of SAs.

So for SA to overrule HA, it should take an extreme majority of SA.

Reply:
Here I still keep proposition as in manifesto for SA vs HA, which is 1:5, because one excess rank gives you 5 more. But promotion changes for A vs SA, and A vs HA, 1:5 and 1:10 respectively which I think is less important.